src/Query.php

```diff
@@ -716,55 +716,50 @@ protected function order(Select $select)
             $orderBy = SortUtil::createOrderBy($defaultSort);
         }
 
-        $columnsAndDirections = [];
         $orderByResolved = [];
         $resolver = $this->getResolver();
         $selectedColumns = $select->getColumns();
 
         foreach ($orderBy as $part) {
             list($column, $direction) = $part;
 
-            if (isset($selectedColumns[$column])) {
+            if (! $column instanceof ExpressionInterface && isset($selectedColumns[$column])) {
                 // If it's a custom alias, we have no other way of knowing it,
                 // unless the caller explicitly uses it in the sort rule.
                 $orderByResolved[] = $part;
             } else {
-                // Prepare flat ORDER BY column(s) and direction(s) for requireAndResolveColumns()
-                $columnsAndDirections[$column] = $direction;
-            }
-        }
-
-        foreach (
-            $resolver->requireAndResolveColumns(array_keys($columnsAndDirections)) as list($model, $alias, $column)
-        ) {
-            $direction = reset($columnsAndDirections);
-            $selectColumns = $resolver->getSelectColumns($model);
-            $tableName = $resolver->getAlias($model);
-
-            if ($column instanceof ExpressionInterface) {
-                if (is_int($alias) && $column instanceof AliasedExpression) {
-                    $alias = $column->getAlias();
-                } elseif (is_string($alias) && $model !== $this->getModel()) {
-                    $alias = $resolver->qualifyColumnAlias($alias, $tableName);
-                }
+                foreach (
+                    $resolver->requireAndResolveColumns([$column]) as list($model, $alias, $column)
```

Comment:
I know you had to adjust the implementation quite a bit, but with this `requireAndResolveColumns` only works with a single column a time, multiple times. I'm not sure about the side-effects this might have. Is there a chance to keep calling it only once?